Mathlib/GroupTheory/Subsemigroup/Center.lean

```diff
@@ -20,79 +20,169 @@ import Mathlib.GroupTheory.Subsemigroup.Operations
 
 We provide `Submonoid.center`, `AddSubmonoid.center`, `Subgroup.center`, `AddSubgroup.center`,
 `Subsemiring.center`, and `Subring.center` in other files.
+
+## References
+
+* [Cabrera García and Rodríguez Palacios, Non-associative normed algebras. Volume 1]
 -/
 
 
 variable {M : Type*}
 
 namespace Set
 
+/-- Conditions for an element to be additively central -/
+structure IsAddCentral [Add M] (z : M) : Prop where
+  /-- addition commutes -/
+  comm (a : M) : z + a = a + z
+  /-- associative property for left addition -/
+  left_assoc (b c : M) : z + (b + c) = (z + b) + c
+  /-- middle associative addition property -/
+  mid_assoc (a c : M) : (a + z) + c = a + (z + c)
+  /-- associative property for right addition -/
+  right_assoc (a b : M) : (a + b) + z = a + (b + z)
+
+/-- Conditions for an element to be multiplicatively central -/
+@[to_additive]
+structure IsMulCentral [Mul M] (z : M) : Prop where
+  /-- multiplication commutes -/
+  comm (a : M): z * a = a * z
+  /-- associative property for left multiplication -/
+  left_assoc (b c : M) : z * (b * c) = (z * b) * c
+  /-- middle associative multiplication property -/
+  mid_assoc (a c : M) : (a * z) * c = a * (z * c)
+  /-- associative property for right multiplication -/
+  right_assoc (a b : M) : (a * b) * z = a * (b * z)
+
 variable (M)
 
 /-- The center of a magma. -/
 @[to_additive addCenter " The center of an additive magma. "]
 def center [Mul M] : Set M :=
-  { z | ∀ m, m * z = z * m }
+   { z | (IsMulCentral z) }
 #align set.center Set.center
 #align set.add_center Set.addCenter
 
 -- porting note: The `to_additive` version used to be `mem_addCenter` without the iff
 @[to_additive mem_addCenter_iff]
-theorem mem_center_iff [Mul M] {z : M} : z ∈ center M ↔ ∀ g, g * z = z * g :=
+theorem mem_center_iff [Mul M] {z : M} : z ∈ center M ↔ IsMulCentral z :=
   Iff.rfl
 #align set.mem_center_iff Set.mem_center_iff
 #align set.mem_add_center Set.mem_addCenter_iff
 
-instance decidableMemCenter [Mul M] [∀ a : M, Decidable <| ∀ b : M, b * a = a * b] :
+variable {M}
+
+@[to_additive]
+theorem _root_.Subsemigroup.mem_center_iff [Semigroup M] {z : M} :
+    z ∈ Set.center M ↔ ∀ g, g * z = z * g := by
+  constructor
+  · exact fun a g ↦ by rw [Set.IsMulCentral.comm a g]
+  · intro h
+    constructor
+    · intro _
+      rw [h]
+    · intros _ _
+      rw [mul_assoc]
+    · intros _ _
+      rw [mul_assoc]
+    · intros _ _
+      rw [mul_assoc]
+#align subsemigroup.mem_center_iff Subsemigroup.mem_center_iff
+#align add_subsemigroup.mem_center_iff AddSubsemigroup.mem_center_iff
+
+variable (M)
+
+instance decidableMemCenter [Mul M] [∀ a : M, Decidable <| IsMulCentral a] :
```

Comment:
This is going to be useless, because there are no `Decidable` instances about `IsMulCentral a`. You should either:
* Leave it for only the `Semigroup` case for now, perhaps with a TODO comment
* Add an `instance : Decidable (IsMulCentral a)` that takes `[∀ a : M, Decidable <| ∀ b : M, b * a = a * b]` and three other decidable arguments

I doubt the second option is worth the effort, but the first option is easy and more useful than what you haev here.

Reply:
Okay, I've done the first option.